geode-gfsh/src/main/java/org/apache/geode/management/internal/cli/functions/ReplicateRegionFunction.java

```diff
@@ -0,0 +1,362 @@
+    }
+    return new CliFunctionResult(context.getMemberName(), CliFunctionResult.StatusState.OK,
+        "Canceled command");
+  }
+
+  public static String getReplicateRegionFunctionThreadName(String regionName, String senderId) {
+    return "replicateRegionFunctionThread_" + regionName + "_" + senderId;
+  }
+
+  /**
+   * If a complete batch in the last cycle has not been replicated yet it returns false.
+   * Otherwise, it returns true and runs the actions to be done when a batch has been
+   * replicated: throw an interrupted exception if the operation was canceled and
+   * adjust the rate of replication by sleeping if necessary.
+   *
+   * @param startTime time at which the entries started to be replicated
+   * @param replicatedEntries number of entries replicated so far
+   * @param maxRate maximum rate of replication
+   */
+  void doActionsIfBatchReplicated(long startTime, int replicatedEntries, long maxRate)
+      throws InterruptedException {
+    if (Thread.currentThread().isInterrupted()) {
+      throw new InterruptedException();
+    }
+    if (maxRate == 0) {
+      return;
+    }
+    final long currTime = clock.millis();
+    final long elapsedMs = currTime - startTime;
+    if (elapsedMs == 0 || replicatedEntries * 1000L / elapsedMs > maxRate) {
+      final long targetElapsedMs = (replicatedEntries * 1000L) / maxRate;
+      final long sleepMs = targetElapsedMs - elapsedMs;
+      logger.info("Sleeping for {} ms", sleepMs);
+      threadSleeper.millis(sleepMs);
+    }
+  }
+
+  private EntryEventImpl createEventForReplicatedRegion(InternalCache cache, InternalRegion region,
+      Region.Entry entry) {
+    return createEvent(cache, region, entry);
+  }
+
+  private EntryEventImpl createEventForPartitionedRegion(GatewaySender sender, InternalCache cache,
+      InternalRegion region,
+      Region.Entry entry) {
+    EntryEventImpl event = createEvent(cache, region, entry);
+    BucketRegion bucketRegion = ((PartitionedRegion) event.getRegion()).getDataStore()
+        .getLocalBucketById(event.getKeyInfo().getBucketId());
+    if (bucketRegion != null && !bucketRegion.getBucketAdvisor().isPrimary()
+        && sender.isParallel()) {
+      return null;
+    }
+    if (bucketRegion != null) {
+      bucketRegion.handleWANEvent(event);
+    }
+    return event;
+  }
+
+  private EntryEventImpl createEvent(InternalCache cache, InternalRegion region,
+      Region.Entry entry) {
+    EntryEventImpl event = new DefaultEntryEventFactory().create(region, Operation.UPDATE,
+        entry.getKey(),
+        entry.getValue(), null, false,
+        (cache).getInternalDistributedSystem().getDistributedMember());
+    if (entry instanceof NonTXEntry) {
+      event.setVersionTag(((NonTXEntry) entry).getRegionEntry().getVersionStamp().asVersionTag());
+    } else {
+      event.setVersionTag(((EntrySnapshot) entry).getVersionTag());
+    }
+    event.setNewEventId(cache.getInternalDistributedSystem());
+    return event;
```

Comment:
Events are bouncing back and forth between sites currently.

You'll have to create a GatewaySenderEventCallbackArgument for the EntryEvent so it knows the originator and recipient sites like:
```
GatewaySenderEventCallbackArgument geCallbackArg = new GatewaySenderEventCallbackArgument(null, cache.getInternalDistributedSystem().getDistributionManager().getDistributedSystemId(), remoteDsIds);
event.setCallbackArgument(geCallbackArg);
```

Reply:
The other thing we talked about is to prevent the callbacks from occurring altogether. If you do that, you won't need to generate the GatewaySenderEventCallbackArgument. Instead, you'll have to modify GatewayReceiverCommand and basicBridgeCreate/Put/Destroy to set generateCallbacks to false:
```
final EntryEventImpl event = entryEventFactory.create(this, Operation.UPDATE, key,
    null, theCallbackArg, false,
    memberId.getDistributedMember(), false, eventID);
```
I guess you'll need to add a parameter to the methods. I think that would mean adding actions to the GatewaySenderEventImpl and processing those actions appropriately in GatewayReceiverCommand.

Reply:
@boglesby I have added a new commit to prevent the callbacks from being fired.
I did not add a new field to the `GatewaySenderEventImpl` to avoid issues with rolling upgrades. Instead, I created a new action in the `GatewaySenderEventImpl` that is used when the source event from which the `GatewaySenderEventImpl` is created has `generateCallbacks` set to false and a new `EnumListenerEvent` value to be passed to the `GatewaySenderEventImpl` to use the value of generateCallbacks from the source event so that, if it is false, the new action (to indicate that generate callbacks should not be done) is used.